Create video-img.py

RuochenYin opened the pull request:
convert video to images

RuochenYin committed:
Create video-img.py

convert video to images

MarcBresson commented on video-img.py:
```diff
@@ -0,0 +1,65 @@
+    if not os.path.exists(os.path.join(result_path, name)):
+        os.makedirs(os.path.join(result_path, name))
+    # 视频到图片 (video to image)
+    save_img(video_path, result_path, interval_num, resize_wh, rot90, show)
```
You can delete this entire section.
`if __name__ == '__main__':` is only when you want to execute a standalone file by invoking the script from the command line 'python video-img.py', but we don't need that.

MarcBresson commented on video-img.py:
```diff
@@ -0,0 +1,65 @@
+    :param resize_wh: 指定保存图片的长宽 (Specifies the length and width of the saved image)
+    :param rot90: 是否90度旋转图片 (Whether to rotate the picture 90 degrees)
+    :param show: 是否进行 窗口展示图片 (Whether to display the picture in the window)
+    :return:
```
delete everything in chinese

MarcBresson commented on video-img.py:
```diff
@@ -0,0 +1,65 @@
+    :param interval_num: 每间隔几帧保存一帧 (Save a frame every few frames)
+    :param resize_wh: 指定保存图片的长宽 (Specifies the length and width of the saved image)
+    :param rot90: 是否90度旋转图片 (Whether to rotate the picture 90 degrees)
+    :param show: 是否进行 窗口展示图片 (Whether to display the picture in the window)
```
delete these two arguments and everything related. We are not going to need this

MarcBresson commented on video-img.py:
```diff
@@ -0,0 +1,65 @@
+    :param video_path: 视频路径
+    :param result_path: 结果保存路径
+    :param interval_num: 每间隔几帧保存一帧 (Save a frame every few frames)
+    :param resize_wh: 指定保存图片的长宽 (Specifies the length and width of the saved image)
```
We will work only with horizontal videos. Video resolution is usually caracterized only by its height (e.g. [1080p)](https://en.wikipedia.org/wiki/1080p). Because of that, I think it would be better to ask for a **target height**.
For instance, if we feed a 368p video to this function, we could specify `target_height=1080` to resize the extracted frames to 1080 pixels in height, and the width calculated so that we have the same aspect ratio in the end.

MarcBresson commented on video-img.py:
```diff
@@ -0,0 +1,65 @@
+
+
+
+def save_img(video_path, result_path, interval_num, resize_wh, rot90=False, show=False):
```
specify the type of each variable. [https://peps.python.org/pep-0484/](https://peps.python.org/pep-0484/) for reference.

MarcBresson commented on video-img.py:
```diff
@@ -0,0 +1,65 @@
+import time
+import numpy as np
+import os
+
```
import Path from [pathlib](https://docs.python.org/3/library/pathlib.html). It helps dealing with path

MarcBresson commented on video-img.py:
```diff
@@ -0,0 +1,65 @@
+    Save each frame of the video to a folder, naming format: video name_timestamp_num
+    :param video_path: 视频路径
+    :param result_path: 结果保存路径
+    :param interval_num: 每间隔几帧保存一帧 (Save a frame every few frames)
```
specify the desired extraction frame rate instead of interval_num

MarcBresson reviewed (changes requested):
Move the file under src/utils

RuochenYin committed:
Create video-img_2.py

convert video to img

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+    print("Totally save {:d} images".format(index-1))
+
+
+def main():
+    # recursively delete the folder where the frame images was stored before,
+    # and create a new one
+    try:
+        shutil.rmtree(EXTRACT_FOLDER)
+    except OSError:
+        pass
+    import os
+    os.mkdir(EXTRACT_FOLDER)
+    # Extract the frames and save it to the specified path
+    extract_frames(VIDEO_PATH, EXTRACT_FOLDER, 1, 25)
```
This is only for your testing. It should not appear in the script

MarcBresson replied:
I will show you how I test my scripts tomorrow morning

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+import shutil
+from pathlib import Path
+
+# global variable
+VIDEO_PATH = Path("D:/GDP/TESTING/nofi001.mp4")
+EXTRACT_FOLDER = Path("D:/GDP/TESTING/vi-im/")
```
This is only for your testing, it should not appear in the script

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+VIDEO_PATH = Path("D:/GDP/TESTING/nofi001.mp4")
+EXTRACT_FOLDER = Path("D:/GDP/TESTING/vi-im/")
+
+def extract_frames(video_path, result_path, index, desired_fps):
```
use typing hints for your variables (you can look up `python typing hints` on google)

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+    '''
+        :param video_path: video path
+        :param result_path: the location where the frame image is stored
+        :param index: the starting sequence number of the frame save filename,
+                      this sequence number will be appended to the filename of each saved frame image
```
- add a default value to the parameter (0 is a good default value)
- move this parameter after `desired_fps` because it is of lower importance (we will often want our extraction to start at index 0)
- this parameter could even be deleted, but we will discuss about it tomorrow :)

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+        :param desired_fps
+        '''
+
+    video = cv2.VideoCapture() # read frames from video source
+    cap = cv2.VideoCapture(str(video_path))
```
Is this normal that you create two different objects of VideoCapture ?

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+    fps = cap.get(cv2.CAP_PROP_FPS) # get the frame rate of the video
+    EXTRACT_FREQUENCY = int(fps / desired_fps) # Frame extraction frequency
+    print("EXTRACT_FREQUENCY: ", EXTRACT_FREQUENCY)
+    if not video.open(str(video_path)):
+        print("can not open the video")
+        exit(1)
```
This should be an error handling (search `raise exceptions python` on google for more info)

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+    cap = cv2.VideoCapture(str(video_path))
+    fps = cap.get(cv2.CAP_PROP_FPS) # get the frame rate of the video
+    EXTRACT_FREQUENCY = int(fps / desired_fps) # Frame extraction frequency
+    print("EXTRACT_FREQUENCY: ", EXTRACT_FREQUENCY)
```
we do not need verbose. It is good for debugging but will flood the actual console when we will be dealing with a lot of videos

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+            index += 1
+        count += 1
+    video.release()
+    print("Original fps: ", fps)
+    print("Totally save {:d} images".format(index-1))
```
verbose (same as the comment on line 22)

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+        if count % EXTRACT_FREQUENCY == 0:
+            # Determine whether the current frame needs to be extracted
+            save_path = "{}/{:>03d}.jpg".format(result_path, index)
+            # save the current frame as an image file
```
Most of the time, comments with `#` are redondant with the code. The code is generally explicit enough to avoid the comments :)
I think you can delete everyone one of them

MarcBresson commented on src/utils/video-img_2.py:
```diff
@@ -0,0 +1,53 @@
+            break
+        if count % EXTRACT_FREQUENCY == 0:
+            # Determine whether the current frame needs to be extracted
+            save_path = "{}/{:>03d}.jpg".format(result_path, index)
```
having a fix length `:>03d` (actual number preceded by 0) is dangerous if for some reason we have more than 999 frames. The number of outputed image can be computed beforehand to allow for a dynamic left padding with zeros. But this is of low priority.

MarcBresson reviewed (changes requested):
rename the file without the 2 because it is not pertinent (you may have a version 1 on your computer, but git already handles the versionning). You could rename it `video_to_images` for instance.

MarcBresson commented:
Two more things:
- You can delete the initial `video-img.py` script as we don't need it anymore
- We are going to use the numpy docstring convention for the project [https://numpydoc.readthedocs.io/en/latest/format.html](https://numpydoc.readthedocs.io/en/latest/format.html)

RuochenYin committed:
video to image

RuochenYin committed:
doc: convert to numpy docstring conventions

In order to have an uniform docstring experience on the project,
the numpy conventions are used.

MarcBresson commented:
Branch merged into dev-dataset-processing
